### ISSUE TITLE
Parametrized Nonlinearities

### ISSUE THREAD

**Issue by ebattenberg** [benanne/Lasagne#133]
[MSR's new work](http://arxiv.org/abs/1502.01852) using parametric rectifiers has renewed my interest in this topic.  Previous work on piece-wise linear units seemed a bit overly complicated for the performance improvement it achieved, but this simpler 2-part linear unit seems like it would be fairly easy to implement, and it helped MSR set the new (new new new...) state-of-the-art on Imagenet.  

The only architectural consideration from a Lasagne standpoint would be the fact that the nonlinearity portion of a layer would have its own parameters.  I know that Maxout is a kind of parametrized nonlinearity where the parameters are the incoming weights (and it can be implemented by max pooling across channels after a linear layer).  And dropout has been implemented as a separate layer.  With these in mind, maybe it would make sense to just make a parametric ReLU just a separate layer rather than one of the simple parameter-free nonlinearity types that come along with existing layers.  

There is probably a lot more pressing working to be done as far as Lasagne is concerned, but I just wanted to make sure we were thinking about this possibility during the planned re-architecting of certain features.

**Comment by benanne** [comment]
I've also been thinking about this, and I agree that this is probably better implemented as a `Layer` subclass, at least for now. If other interesting use cases for parameterised nonlinearities come up we can consider it, but I have a feeling they're going to add a lot of complexity (`get_all_params` would have to be modified to inspect layers' nonlinearities for example).

**Comment by diogo149** [comment]
With the new Batch Normalization paper out, I've decided to try all of these alternative activations and do a comparison (and this could address https://github.com/benanne/Lasagne/issues/85 as well). (:

My current idea for an implementation would be to have all nonlinearities as layers for consistency (kind of like in caffe). It seems like it should be possible to have them both as a function and as a layer if backwards compatibility / syntax is an issue, though that violates Python's "There should be one-- and preferably only one --obvious way to do it."

I'd happily make the necessary changes (or learn why it's a bad idea).

**Comment by benanne** [comment]
The current approach of having (simple) nonlinearities be part of a layer, with the rectifier as the default, is mainly because that's the most common use case. It's one of those issues where I think pragmatism trumps 'cleanliness' of the design. And it also makes it a bit easier to use custom nonlinearities. I like the 'sensible defaults' approach and I think rectification is the most sensible default thing to do (at least for now).

But for nonlinearities that have their own parameters, this approach breaks down, so I would definitely implement those as layers. And then maybe there is something to be said for implementing all the others as layers as well for consistency... tough problem :) We should definitely discuss it.

**Comment by diogo149** [comment]
I think a nice middleground would be to have nonlinearities as both functions and layers, and just have the default nonlinearity identity. It sacrifices a small amount of cleanliness (specifying `nonlinearity=relu`) for more clarity. (:

Anyone have thoughts on `ReLU(BatchNormalization(ConvLayer(nonlinearity=identity)))` vs `BatchNormalization(ConvLayer(nonlinearity=identity), nonlinearity=relu)`?

**Comment by f0k** [comment]
> have the default nonlinearity identity

I'd also like that.

> And then maybe there is something to be said for implementing all the others as layers as well for consistency...

I think having nonlinearities as part of a layer is more practical. But we could add a layer that does nothing but apply a nonlinearity, a) because it allows people to have a distinct nonlinearity layer, and b) because `nonlinearity(maxpool(conv))` is a tad faster than `maxpool(nonlinearity(conv))`, but it's currently not possible to express the former with Lasagne.

**Comment by craffel** [comment]
>> have the default nonlinearity identity

> I'd also like that.

:+1: more

**Comment by skaae** [comment]
I have tried to implement the batch normalization method from http://arxiv.org/abs/1502.03167

Its a little hacky though. I'm not sure that the code is correct, but it seems converge. 

I'm unsure about the running average mentioned in the article is supposed to be implemented. An also about how the mean and variance values are supposed to be calculated during inference. 

Batch normalization 
https://gist.github.com/skaae/5faacedb9c5961136e82

MNIST code
https://gist.github.com/skaae/8c7cca5e4b9621487738

**Comment by benanne** [comment]
Looks like you're dividing by the variance, I think you want to divide by the square root of that instead.

**Comment by skaae** [comment]
woops, yes :)

**Comment by skaae** [comment]
btw when the article says

"""
For convolutional layers, we additionally want the nor-
malization to obey the convolutional property – so that
different elements of the same feature map, at different
locations, are normalized in the same way. To achieve
this, we jointly normalize all the activations in a mini-
batch, over all locations.
"""

And output from a Conv2D is (batch_size, numfilters, width, height). 
Should i then share mean and variance for each filter. 
So if have say 32 filters i would have 32 means and 32 variances?

**Comment by benanne** [comment]
Yep, that's what it means. You basically just need to broadcast across the width and height dimensions.

**Comment by ebattenberg** [comment]
Maybe move this to a new thread? I'm interested in this topic as well but
it doesn't really fit under "parametrized nonlinearities". :)

**Comment by aloisg** [comment]
I'm not convinced you should apply Batch Normalization for the softmax layer in your MNIST tutorial.

"The last hidden layer is followed by a fully-connected layer with 10 activations (one per class) and cross-entropy loss", and then "We added Batch Normalization to each hidden layer of the network".

Tell me what you think about that, maybe i'm wrong.

**Comment by skaae** [comment]
I think you are right

**Comment by ebattenberg** [comment]
> Maybe move this to a new thread? I'm interested in this topic as well but
> it doesn't really fit under "parametrized nonlinearities". :)

@skaae What I meant here is that you should start a new issue for "Batch Normalization".  It's a big enough topic by itself. :)

**Comment by skaae** [comment]
Will do :) Batch normalization Moved to https://github.com/benanne/Lasagne/issues/141

**Comment by instagibbs** [comment]
Accidentally started a new PR for this old topic. I agree with others in thinking adding all non-linearities as additional layers is the way forward, as long as default for Dense/Conv/etc layers are identity. I still think the primary/advertised way should be one or the other. The proposed new way is much cleaner, imo, and less confusing to new users.  

Is anyone working on this?

**Comment by f0k** [comment]
> I still think the primary/advertised way should be one or the other.

I'd advertise `nonlinearity=...` as the primary way and note that the other one is possible as well, and allows for parameterized, learnable nonlinearities. Keras does it quite similarly in their README.

> The proposed new way is much cleaner, imo, and less confusing to new users.

Which "new way" do you refer to here? Having nonlinearities as separate layers?

> Is anyone working on this?

We already added a layer that does nothing but apply a nonlinearity, that was in #225.
We more or less agreed on changing the default nonlinearity to `identity` in a dev meeting, but there are still objections within the core team and nobody started working on it. I guess the more arguments or proponents we have for changing it, the more likely it will happen. (Full disclosure: I'm one of the proponents.) Discussion on that should take place in #138.

**Comment by instagibbs** [comment]
I was referring to non-linearities as layers, yes.  

Has there been any discussion on how to extend NonlinearityLayer to a parameterized version? 

I suppose now that the nonlinearity is encapsulated in its own layer additional parameters can be done as special cases inside that layer. Another alternative is yet another subclass for (specific?) parameterized versions.

**Comment by benanne** [comment]
I'm still firmly in favour of keeping rectification as the default, but if the majority votes for linear layers by default instead I will reluctantly surrender :) It's going to make all my code a lot more verbose, that's for sure...

I don't think non-linearities should be layers unless they absolutely have to be. The notion of a 'layer' in neural network research almost always includes the nonlinearity, and I think we shouldn't stray from that unless we have to. (This discussion has come up before a few times.) It creates a mismatch between the description of a model and its actual implementation. I like that this is currently an almost 1-to-1 mapping in Lasagne (not taking into account things like `ConcatLayer`, but I think that's a necessary evil).

Of course this breaks down when the nonlinearities suddenly have learnable parameters of their own, which is why I think we should implement PReLU as a layer class (there is no reason why this should extend `NonlinearityLayer` by the way). There is no point in creating a new abstraction specifically for this use case as the `Layer` abstraction already handles this reasonably well.

> I suppose now that the nonlinearity is encapsulated in its own layer 

Just to clarify, this layer exists to cater to certain specialised use cases and is definitely not intended to be 'the' way to add nonlinearities in your Lasagne nets. The default way is to use the `nonlinearity` keyword argument. `NonlinearityLayer` should only be used if that doesn't cut it.

**Comment by instagibbs** [comment]
Makes sense. I'll probably take a crack at it then.

**Comment by f0k** [comment]
> It's going to make all my code a lot more verbose, that's for sure...

More explicit! It will make all your code more explicit! :)

> Makes sense. I'll probably take a crack at it then.

Just to guide you, the solution would be something like:
```python
class PReLULayer(Layer):
    def __init__(self, alpha=0.25, **kwargs):
        super(PReLULayer, self).__init__(**kwargs)
        self.alpha = self.create_param(alpha, ())
    def get_output_for(self, input):
        return LeakyRectify(self.alpha)(input)
    def get_params(self):
        return [self.alpha]
```
Plus documentation and tests. There are some things in Lasagne that don't play well with scalar parameters, though, we'd either need to change that or have the shape of `alpha` be `(1,)` instead (and possibly unwrap it into a scalar in `get_output_for` so it broadcasts correctly).

**Comment by instagibbs** [comment]
I stuck with using scalars for the first attempt, and the only place I had to change code to support it appears to be the base Layer class in create_param.  

Are there other places that you know are problematic?

**Comment by benanne** [comment]
What changes were they? `create_param` is currently being revamped / renamed to `add_param` in #228 so it would be nice to make it work properly for scalars in one go :)

**Comment by f0k** [comment]
> Are there other places that you know are problematic?

I haven't actually checked, but off the top of my head I assume `lasagne.layers.count_params` will need to be adapted as well, and possibly the routines for getting/setting parameter values (as numpy arrays of shape `()` tend to get converted to Python floats sometimes). `lasagne.utils.floatX` could also be changed to support scalars; I'll do a PR for that.

**Comment by instagibbs** [comment]
>What changes were they?

Just embedded another if case in the type checks:

>elif isinstance(param, np.float32):
>            return theano.shared(param, name=name)

@f0k I'll change up count_params as well, thanks for the tip!
